# Inferential Statistics:

Inferential statistics is used to gather some information about the population from the sample of data collected.<br>
There are many tests which are used in Inferential statistics depending on the objective and the data.<br>

1. To find out ; Is there a significant difference in the marks of students between section A and section B. For any objective, we formulate a hypothesis.<br><br>

Null Hypothesis (Ho):<br>
There is no significant difference in the marks between students of section A and section B.<br><br>

Alternative Hypothesis (Ha):<br>
There is a significant difference in the marks of section A and section B.<br><br>


# The Test to be used:

Step 1: Check the normality separately for the variables Wing A and Wing B.<br>
Step 2: If both the variables have **normal distribution**, then we use a parametric test known as **independent sample t-test**.<br>
Step 3: If the variables don't have a **normal distribution** then we use a non-parametric test known as **Mann Whitney U test**.<br>

## How to do Independent sample t-test in Python?
import the library and inport the dataset into Python.<br>

To do the independent sample t-test in Python, the function used is **ttest_ind**.<br>

### How to import a specific function from a library
To import the function **ttest_ind** from the Python library **scipy.stats**.

In [1]:
from scipy.stats import ttest_ind

### How to implement "independent sample t-test"?

There are two variables **Wing A** and **Wing B**<br>
The independent sample test will give a "t" value and this **t-value** has to be compared with the **t-critical** value<br>
The code to implement the independent sample t-test:

In [3]:
import pandas as pd
df = pd.read_csv('D:\students.csv')
df

Wing A  Wing B
0    10.70    7.20
1     9.89    6.68
2    11.83    9.29
3     9.04    8.95
4     9.37    6.61
5    11.68    8.53
6     8.36    8.92
7     9.76    7.95
8    13.67    7.57
9     8.96    6.38
10    9.51    8.89
11   10.85   10.03
12   10.57    9.30
13   11.06    5.28
14    8.91    9.23
15   11.79    9.25
16   10.59    8.44
17    9.13    6.57
18   12.37   10.61
19    9.91    6.77

In [5]:
t_statistics, p_value = ttest_ind( df['Wing A'], df['Wing B'])
t_statistics

5.16151166403543

The value from Python we get is the **t-calculated**, this value has to be compared with the **critical value**. This **critical value** will be obtained from t-table.

Use the column for 5 % significance level.<br>
The **dof**= n-2, where **n** is the sample size.<br>
Get the critical value from the statistical table.<br>

### The Interpretation
If the calculated from Python is less than the critical value from the table, then you accept the Ho hypothesis  and conclude there is no difference between the two sections.<br><br>

If the calculated value from Python is greater than the critical from the table, then the Ha is accepted and we conclude there is a difference in the performance of the students between the two sections.<br>

# Checking the Normality for "Wing A"

In [6]:
df.describe()

Wing A     Wing B
count  20.000000  20.000000
mean   10.397500   8.122500
std     1.370032   1.417195
min     8.360000   5.280000
25%     9.310000   6.747500
50%    10.240000   8.485000
75%    11.215000   9.235000
max    13.670000  10.610000

## Criteria 1:

In [8]:
freq_table = pd.crosstab(df['Wing A'],'count')
freq_table

col_0   count
Wing A       
8.36        1
8.91        1
8.96        1
9.04        1
9.13        1
9.37        1
9.51        1
9.76        1
9.89        1
9.91        1
10.57       1
10.59       1
10.70       1
10.85       1
11.06       1
11.68       1
11.79       1
11.83       1
12.37       1
13.67       1

In [9]:
x1 = 10.397500 - 1.370032
x2 = 10.397500 + 1.370032

print("x1=",x1)
print("x2=",x2)

x1= 9.027468
x2= 11.767532000000001


**The counted values are 13, therefore the area is <br>
Area = 13/20 = 0.65**

## Criteria 2:

In [11]:
freq_table = pd.crosstab(df['Wing A'],'count')
freq_table

col_0   count
Wing A       
8.36        1
8.91        1
8.96        1
9.04        1
9.13        1
9.37        1
9.51        1
9.76        1
9.89        1
9.91        1
10.57       1
10.59       1
10.70       1
10.85       1
11.06       1
11.68       1
11.79       1
11.83       1
12.37       1
13.67       1

In [12]:
x1 = 10.397500 - 2*1.370032
x2 = 10.397500 + 2*1.370032

print("x1=",x1)
print("x2=",x2)

x1= 7.657436000000001
x2= 13.137564000000001


**The counted values are 19, therefore the area is <br>\
Area = 19/20 = 0.95**


## Criteria 2:

In [13]:
freq_table = pd.crosstab(df['Wing A'],'count')
freq_table

col_0   count
Wing A       
8.36        1
8.91        1
8.96        1
9.04        1
9.13        1
9.37        1
9.51        1
9.76        1
9.89        1
9.91        1
10.57       1
10.59       1
10.70       1
10.85       1
11.06       1
11.68       1
11.79       1
11.83       1
12.37       1
13.67       1

In [14]:
x1 = 10.397500 - 3*1.370032
x2 = 10.397500 + 3*1.370032

print("x1=",x1)
print("x2=",x2)

x1= 6.287404000000001
x2= 14.507596


**The counted values are 20, therefore the area is <br>
Area = 20/20 = 1**

## Conclusion:
The Wing A variable is nearly normally distributed.

# Checking the Normality for Wing B

In [17]:
df.describe()

Wing A     Wing B
count  20.000000  20.000000
mean   10.397500   8.122500
std     1.370032   1.417195
min     8.360000   5.280000
25%     9.310000   6.747500
50%    10.240000   8.485000
75%    11.215000   9.235000
max    13.670000  10.610000

## Criteria 1:

In [18]:
freq_table = pd.crosstab(df['Wing B'],'count')
freq_table

col_0   count
Wing B       
5.28        1
6.38        1
6.57        1
6.61        1
6.68        1
6.77        1
7.20        1
7.57        1
7.95        1
8.44        1
8.53        1
8.89        1
8.92        1
8.95        1
9.23        1
9.25        1
9.29        1
9.30        1
10.03       1
10.61       1

In [19]:
x1 = 8.122500 - 1.417195
x2 = 8.122500 + 1.417195

print("x1=",x1)
print("x2=",x2)

x1= 6.705305000000001
x2= 9.539695


**The counted values are 13, therefore the area is <br>
Area = 13/20 = 0.65**

## Criteria 2:

In [ ]:
freq_table = pd.crosstab(df['Wing B'],'count')
freq_table

In [1]:
x1 = 8.122500 - 2*1.417195
x2 = 8.122500 + 2*1.417195

print("x1=",x1)
print("x2=",x2)

x1= 5.2881100000000005
x2= 10.956890000000001


## Criteria 3:

In [ ]:
freq_table = pd.crosstab(df['Wing B'],'count')
freq_table

In [2]:
x1 = 8.122500 - 3*1.417195
x2 = 8.122500 + 3*1.417195

print("x1=",x1)
print("x2=",x2)

x1= 3.870915
x2= 12.374085000000001


# t-test

## Looking for "t-value" from Python :

In [15]:
t_statistics, p_value = ttest_ind( df['Wing A'], df['Wing B'])
t_statistics

5.16151166403543

The t-value from Python is 5.162 , and the critical value from the table is 2.101, 
therefore the alternate hypothesis is accepted which means threre is a significant difference in the performance of students 
between the two sections.

In [ ]:
## How to create a dataframe with Python
## Mann Whitney U test
## paired sample t-test

# Creating DataFrame in Python

## How to create a DataFrame in Python?

**If an external file is not provided then we need to create the dataframe directly in Python.**

For creating a dataframe in python for the variable **Wing A**.

In [2]:
import pandas as pd

In [3]:
data = [1,2,3,4,20] # actual data


Using this object **data** , we are trying to create a data frame **df**.

In [4]:
df = pd.DataFrame(data,columns=['Wing A'])
df

Wing A
0       1
1       2
2       3
3       4
4      20

## Q-> Create a dataframe df which consists a single variable called Wing A


In [6]:
data = [10.7,9.89,11.83,9.04,9.37,11.68,8.36,9.76,13.67,8.96,9.51,10.85,10.57,11.06,8.91,11.79,10.59,9.13,12.37,9.91]

In [8]:
df1 = pd.DataFrame(data,columns=['Wing A'])
df1

Wing A
0    10.70
1     9.89
2    11.83
3     9.04
4     9.37
5    11.68
6     8.36
7     9.76
8    13.67
9     8.96
10    9.51
11   10.85
12   10.57
13   11.06
14    8.91
15   11.79
16   10.59
17    9.13
18   12.37
19    9.91

## Q-> Calculate the mean of Wing A and crosscheck with the mean of the imported file


In [9]:
df1.mean()

Wing A    10.3975
dtype: float64

## Q-> Now create a dataframe with two variables "Wing A" and "Wing B" , and then calculate the mean of the two varibles and cross check with the mean of the imported file.

In [12]:
data2 = [[10.57,7.2],[9.89,6.68],[11.83,9.29],[9.04,8.95],[9.37,6.61],[11.68,8.53],[8.36,8.92],[9.76,7.95],[13.67,7.57],[8.96,6.38],[9.51,8.89],[10.85,10.03],[10.57,9.3],[11.06,5.28],[8.91,9.23],[11.79,9.25],[10.59,8.44],[9.13,6.57],[12.37,10.61],[9.91,6.77]]

# Paired Sample t-test

For doing the **paired sample t-test**,the **students(from DSAT students example)** shouuld remain the same for both the variables.<br>
The function to be used is **ttest_rel**.

Import the function from **scipy.stats**

In [14]:
from scipy.stats import ttest_rel

The **dof** is **n-1**.

## When this test can be done?

**Assumption for doing the Paired Sample t-test:**<br>
Create a new variable *hel* which is difference of var1 and var2 , and then check for its **Normality**.<br>
If this variable has a **normal distribution**, then use **paired sample t-test**, otherwise use a non-parametric test **Wilcoxin Test**.<br>

Step1: Create a dataframe from the data.<br>
Step2: Create a new variable **diff** which is the difference between the two variables.<br>
Step3: Check the normality for this new variable.<br>
Step4: Conduct the Paired Sample t-test, and write your interpretation using the statistical table.<br>

**From the complete dataset, use the 20 rows !!**

In [15]:
data = [[1,2.83,3.505],[2,3.295,3.43],[3,2.71,3.67],[4,2.855,3.355],[5,2.98,3.985],[6,3.065,3.63],[7,3.765,4.57],[8,3.265,3.7],[9,3.17,3.66],[10,2.895,3.25],[11,2.63,2.85],[12,2.83,3.34],[13,2.935,3.63],[14,3.115,3.675],[15,2.985,3.475],[16,3.135,3.605],[17,2.75,3.25],[18,3.205,3.54],[19,3,4.005],[20,3.035,3.595]]

In [18]:
df2 = pd.DataFrame(data,columns = ['Sample','TwoDays','SevenDays'])
df2

Sample  TwoDays  SevenDays
0        1    2.830      3.505
1        2    3.295      3.430
2        3    2.710      3.670
3        4    2.855      3.355
4        5    2.980      3.985
5        6    3.065      3.630
6        7    3.765      4.570
7        8    3.265      3.700
8        9    3.170      3.660
9       10    2.895      3.250
10      11    2.630      2.850
11      12    2.830      3.340
12      13    2.935      3.630
13      14    3.115      3.675
14      15    2.985      3.475
15      16    3.135      3.605
16      17    2.750      3.250
17      18    3.205      3.540
18      19    3.000      4.005
19      20    3.035      3.595

In [19]:
df2['diff'] = df2['TwoDays'] - df2['SevenDays']
df2

Sample  TwoDays  SevenDays   diff
0        1    2.830      3.505 -0.675
1        2    3.295      3.430 -0.135
2        3    2.710      3.670 -0.960
3        4    2.855      3.355 -0.500
4        5    2.980      3.985 -1.005
5        6    3.065      3.630 -0.565
6        7    3.765      4.570 -0.805
7        8    3.265      3.700 -0.435
8        9    3.170      3.660 -0.490
9       10    2.895      3.250 -0.355
10      11    2.630      2.850 -0.220
11      12    2.830      3.340 -0.510
12      13    2.935      3.630 -0.695
13      14    3.115      3.675 -0.560
14      15    2.985      3.475 -0.490
15      16    3.135      3.605 -0.470
16      17    2.750      3.250 -0.500
17      18    3.205      3.540 -0.335
18      19    3.000      4.005 -1.005
19      20    3.035      3.595 -0.560

## Checking the Normality of "diff"

In [20]:
df2.describe()

Sample    TwoDays  SevenDays      diff
count  20.00000  20.000000  20.000000  20.00000
mean   10.50000   3.022500   3.586000  -0.56350
std     5.91608   0.253229   0.345458   0.23835
min     1.00000   2.630000   2.850000  -1.00500
25%     5.75000   2.848750   3.411250  -0.68000
50%    10.50000   2.992500   3.600000  -0.50500
75%    15.25000   3.143750   3.671250  -0.46125
max    20.00000   3.765000   4.570000  -0.13500

In [22]:
freq_1 = pd.crosstab(df2['diff'],'count')
freq_1

col_0   count
diff         
-1.005      2
-0.960      1
-0.805      1
-0.695      1
-0.675      1
-0.565      1
-0.560      1
-0.560      1
-0.510      1
-0.500      2
-0.490      2
-0.470      1
-0.435      1
-0.355      1
-0.335      1
-0.220      1
-0.135      1

## Criteria 1:

In [21]:
x1 = -0.56350 - 0.23835
x2 = -0.56350 + 0.23835
print("x1=",x1)
print("x2=",x2)

x1= -0.80185
x2= -0.32515


**The counted values are 14, therefore the area is<br>
Area = 14/20 = 0.70**

## Criteria 2:

In [24]:
x1 = -0.56350 - 2*0.23835
x2 = -0.56350 + 2*0.23835
print("x1=",x1)
print("x2=",x2)

x1= -1.0402
x2= -0.08679999999999999


**The counted values are 20, therefore the area is <br>
Area = 20/20 = 1**

## Criteria 3:

In [28]:
x1 = -0.56350 - 3*0.23835
x2 = -0.56350 + 3*0.23835
print("x1=",x1)
print("x2=",x2)

x1= -1.27855
x2= 0.15154999999999996


**The ciunted values are 20, therefore the area is<br>
Area = 20/20 = 1**

# Paired t-test

In [27]:
t_statistics, p_value = ttest_rel(df2['TwoDays'],df2['SevenDays'])
t_statistics

-10.572884939003538

The t-value obtained from python is 10.57 (**don't take the sign**), and the critical value from the statistical table for dof=19 is 2.09. Therefpre the Ho hypothesis is rejected and the teaching has been effective.

Q-> Use the file hsb2.csv and do the paired sample t-test with the interpretation for following combinations of variables<br>
- **read** and **write**
- **math** and **science**
- **science** and **socst**

In [29]:
df2 = pd.read_csv("D:\hsb2.csv")
df2

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst
0     70       0     4    1       1     1    57     52    41       47     57
1    121       1     4    2       1     3    68     59    53       63     61
2     86       0     4    3       1     1    44     33    54       58     31
3    141       0     4    3       1     3    63     44    47       53     56
4    172       0     4    2       1     2    47     52    57       53     61
..   ...     ...   ...  ...     ...   ...   ...    ...   ...      ...    ...
195   31       1     2    2       2     1    55     59    52       42     56
196  145       1     4    2       1     3    42     46    38       36     46
197  187       1     4    2       2     1    57     41    57       55     52
198  118       1     4    2       1     1    55     62    58       58     61
199  137       1     4    3       1     2    63     65    65       53     61

[200 rows x 11 columns]

## for "read" and "write"

In [30]:
t_statistics, p_value = ttest_rel(df2['read'],df2['write'])
t_statistics

-0.8673065458794775